[#4] feat: 회원가입

NNIIE opened the pull request:
dosc: read.me

### 💬 어떤 작업인가요? 
> 어떤기능을 / 무엇을 위해 진행된 작업인지
- 회원가입 기능
- user exception

### 📝 작업한 내용을 간략히 설명해주세요.
- 회원가입과 예외처리를 간단하게 구현했습니다.

### 🙋🏻‍♀️ 리뷰 요구사항(선택)
> 리뷰어가 특별이 봐주었으면 하는 부분이 있다면 작성해주세요.
- 텀이 너무 길어지는거 같아 PR 먼저 신청 드립니다.
- 추후 디비연결, 테스트코드 작성 예정입니다.

NNIIE committed:
feat: 회원가입 api, user exception
dosc: read.me

f-lab-ben commented on src/main/java/com/gamemarket/common/entity/BaseEntity.java:
```diff
@@ -0,0 +1,29 @@
+
+@MappedSuperclass
+@Getter
+@EntityListeners(AuditingEntityListener.class)
```
entity listener 어노테이션만 달아놓으면 작동하진 않을거에요~

관련해서 jpa config가 있어야할 듯 합니다 :)

f-lab-ben commented on src/main/java/com/gamemarket/common/entity/BaseEntity.java:
```diff
@@ -0,0 +1,29 @@
+
+    @CreatedDate
+    @Column(nullable = false, updatable = false)
+    private LocalDateTime createAt;
```
일반적으로 createdAt, updatedAt 같이 과거형으로 많이 써요

과거의 시간을 보는 개념이기 때문에 그렇습니다 👍

NNIIE replied:
네이밍이 그편이 훨씬 명확한거 같습니다..! 👍

f-lab-ben commented on src/main/java/com/gamemarket/common/exception/user/UserExceptionCode.java:
```diff
@@ -0,0 +1,22 @@
+@Getter
+public enum UserExceptionCode {
+
+    EXISTS_USER_EMAIL(HttpStatus.BAD_REQUEST, "EXISTS_EMAIL",  "이미 존재하는 이메일 입니다."),
```
code는 int 형이 좋지 않을까요?

FE가 이걸 가지고 분기한다고 가정해보면 어떤게 더 나은지 알 수 있을거에요

예컨데, 이메일 관련 exception이 -10000 ~ -19999 까지라고 하면

fe에서는 그 range로 exception을 분기처리 할 수 있겠죠?

하지만 string의 경우는 일일히 다 매핑을 해줘야해서 불편할거에요

NNIIE replied:
String이어야만 하는 이유가 없다면 fe에서 range별로 분기처리할 수 있는 int 형으로 하는게 맞는거 같습니다 👍

f-lab-ben commented on src/main/java/com/gamemarket/common/exception/user/UserExceptionHandler.java:
```diff
@@ -0,0 +1,21 @@
+public class UserExceptionHandler {
+
+    @ExceptionHandler(UserException.class)
+    public ResponseEntity<String> userExceptionHandler(final UserException exception, final HttpServletRequest request) {
```
ResponseEntity에 담아질 Exception 관련 Reponse가 따로 있으면 좋겠네요

지금은 message만 담아서 보내는데 code까지 보내면 fe에서 분기처리하기 좋을거에요
code : 분기처리 목적, message : 디버깅 용도

GameMarketExceptionResponse 등과같이 따로 클래스 정의해두면 좋습니다

f-lab-ben commented on src/main/java/com/gamemarket/user/application/UserService.java:
```diff
@@ -0,0 +1,43 @@
+        userRepository.save(user);
+    }
+
+    private void validateDuplicateUser(final UserSignUpRequest request) {
```
constoller에서 받은 request를 검증하는 롤은

서비스가 좋을까요, 아니면 컨트롤러가 나을까요?

NNIIE replied:
request에 대한 검증은 controller에서 하는게 역할에 맞습니다.
유저 중복 검증은 각각 서비스마다 정책이 다르고 또한 디비까지 갔다와야 하기 때문에 
service layer의 역할이라고 판단했는데 보통 controller 에서 하나요?

f-lab-ben replied:
디비까지 갔다오지 않아도 되는 경우도 많고(parameter 검증 등)

디비를 다녀오더라도 큰 문제가 되지 않는 경우도 있습니다
JPA 를 사용하면 1차 캐시의 영역이기 때문에
성능상 문제점도 없습니다

JPA 관련된 내용 중 1차캐시만 좀 찾아보고 코멘트로 남겨주세요~

NNIIE replied:
## 영속성 컨텍스트
- 어플리케이션과 데이터베이스 사이에 객체를 보관하는 논리적 개념
- 앤티티매니저를 통해 영속성 컨텍스트에 접근한다.
- 엔티티매니저가 생성되면 영속성 컨텍스트가 1:1로 생성된다.
- 영속성 컨텍스트에서 관리되는 엔티티의 상태를 영속 상태 라고 부른다.

## 1차 캐시
- 영속성 컨텍스트 내부에 엔티티를 보관하는 저장소
- 엔티티 매니저로 조회하거나 변경하는 모든 엔티티는 1차 캐시에 저장된다.
- 트랜잭션의 커밋 / 플러시를 호출하면 1차 캐시에 있는 엔티티의 변경 내역을 데이터베이스에 동기화한다
- 같은 엔티티가 있으면 객체 동일성을 보장한다.
- 트랜잭션의 라이프사이클 동안 유지된다.
- 동작방식
  - 1차 캐시 안에 데이터가 있으면 데이터를 반환한다.
  - 없을 경우 DB에서 조회한 데이터를 1차 캐시에 보관 후(영속상태로 만드는 과정) 1차 캐시의 데이터를 반환한다.

## 2차 캐시
- 어플리케이션 범위의 캐시로, 공유 캐시 라고도 부른다.
- 어플리케이션의 라이프사이클 동안 유지된다.
- 캐시 미스 개념이 적용될 수 있다.
- 캐시된 데이터를 직접 반환하지 않고, 복사본을 만들어 반환한다.
- 영속성 컨텍스트가 다를경우, 동등성은 보장하지만 동일성을 보장하지 않는다.
- 동작방식
  - 2차 캐시 안에 데이터가 있으면 데이터의 복사본을 반환한다.
  - 없을 경우 DB에서 조회한 데이터를 2차 캐시에 보관 후 복사본을 반환한다.

f-lab-ben commented on src/main/java/com/gamemarket/user/domain/entity/User.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private long id;
```
reference type으로 바꿔야 할 것 같아요

이유에 대해서는 primitive type을 쓰면 혼용되서 오해가 빚어지는 부분에 대해 찾아보시면 좋습니다~

f-lab-ben commented on src/main/java/com/gamemarket/user/infra/UserRepository.java:
```diff
@@ -0,0 +1,14 @@
+@Repository
+public interface UserRepository extends JpaRepository<User, Long> {
+
+    boolean existsByNickname(String nickname);
```
return type primitive 인지?

f-lab-ben commented on src/main/java/com/gamemarket/user/application/UserService.java:
```diff
@@ -0,0 +1,43 @@
+        final User user = User.builder()
+                .email(request.getEmail())
+                .nickname(request.getNickname())
+                .password(passwordEncoder.encode(request.getPassword()))
```
passwordEncode 
* encode() 
* 결과가 항상 같을까?
* 아니라면, 이거를 어떻게 같은지를 비교할 수 있는지?
* match()

NNIIE replied:
- `encode()`
  - 바이트단위의 임의의 문자열(salt)을 추가해 해싱하기 때문에 결과는 항상 다르다.
- `matches()`
  - 기존의 해싱된 비밀번호에서 salt를 추출하여 사용자가 입력한 비밀번호와 동일한 해시함수를 사용해 해싱하기 때문에 두개의 값을 비교할 수 있다.

f-lab-ben commented on src/main/java/com/gamemarket/common/config/SecurityConfig.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Bean
+    public BCryptPasswordEncoder bCryptPasswordEncoder() {
+        return new BCryptPasswordEncoder();
```
SCryptEncoder
BCryptEncoder

... 차이점은?

f-lab-ben replied:
salt <<<

f-lab-ben replied:
dldydtjr12!
--> bmladsml@#sdklz
--> dldydtjr12!

match()

NNIIE replied:
- `BCryptPasswordEncoder`
  - 작업요소 라는 설정을 통해 해싱과정을 더 느리게 만들 수 있다.
  - 공격자가 무작위 공격을 시도할 때 필요한 시간을 증가시킴으로써 보안을 강화한다.
  - 즉, 비밀번호를 해싱할때 오래 걸리게 만든다.
- `SCryptEncoder`
  - 해싱과정에 많은 메모리를 사용한다.
  - 공격자가 대규모 해킹장비를 사용할 때, 해시를 빠르게 계산하는것을 어렵게 만들어 보안을 강화한다.
  - 즉, 비밀번호를 해싱할때 많은 메모리를 사용해 여러 프로세스가 동시에 많은 해시를 계산하는것을 어렵게 만든다.

NNIIE committed:
refactor: exception code int 로 수정, exception response 생성

NNIIE committed:
feat: h2 dababase 연동

NNIIE committed:
feat: swagger 추가

NNIIE committed:
refactor: jdbc template 적용

NNIIE committed:
refactor: test code

NNIIE commented:
@f-lab-ben 다시 PR 요청 드립니다..!
- exception
  - response 생성
  - 코드 타입 변경 (String -> int)
- swagger 추가
- h2 database 연동
- jdbc template 적용
- 테스트 코드 작성

f-lab-ben commented on src/main/java/com/gamemarket/common/config/DatabaseConfig.java:
```diff
@@ -0,0 +1,29 @@
+        dataSource.setUsername("sa");
+        dataSource.setPassword("");
+        dataSource.setDriverClassName("org.h2.Driver");
+        dataSource.setUrl("jdbc:h2:tcp://localhost/~/database");
```
jdbc 관련 설정은 property 로 뺴는게 좋아보여요~

f-lab-ben commented on src/main/java/com/gamemarket/common/exception/user/UserExceptionHandler.java:
```diff
@@ -0,0 +1,27 @@
+                .message(ex.getCode().getMessage())
+                .build();
+
+        log.info("[UserException] URI: {}, CODE: {}", request.getRequestURI(), response.getCode());
```
exception 발생 시 log level은 info는 적합하지 않을거 같네요 🤔

f-lab-ben commented on src/main/java/com/gamemarket/user/ui/UserController.java:
```diff
@@ -0,0 +1,29 @@
+
+    private final UserService userService;
+
+    @PostMapping("/user/sign-up")
```
이렇게되면 컨트롤러별 role을 명확하게 분리해서 보기 힘들거에요

UserController면 user관련된 api만 매핑하면 되겠죠? 

그렇다면 클래스 위에 RequestMapping으로 /user 를 잡아주고

postMapping으로 /sign-up 을 잡아준다면 더 명확하게 계층을 구분해낼 수 있습니다

f-lab-ben commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,12 @@
+spring:
+  datasource:
```
java config에 있는 내용과 중복인것 같아서
code에 있는 내용은 지워도 될거같죠?

NNIIE committed:
refactor:
- @ConfigurationProperties, @RequestMapping 사용
- 로그레벨 변경
- 중복확인에 대한 롤 변경 service -> controller

NNIIE commented:
@f-lab-ben 
피드백 적용 후 PR 요청 드립니다..!
다만 유저 중복 검증에 대한 롤을 Controller 로 옮겼는데
제가 피드백을 잘 이해한게 맞는지 궁금합니다!

NNIIE committed:
dosc: read.me

f-lab-ben commented on src/test/java/com/gamemarket/user/ui/UserControllerTest.java:
```diff
@@ -0,0 +1,88 @@
+        request.put("nickname", "abcd");
+        request.put("password", "qwer1234QW!");
+
+        String json = new Gson().toJson(request);
```
Gson을 직렬화 라이브러리로 사용하게 되면
spring 전반에서 의도치 않은 사이드 이펙이 발생할 수 있습니다

예컨데, String to Json 이나 그 반대로 수행해야 하는 작업의 경우

spring에서는 어떤 라이브러리를 사용하고 있고, 똑같이 직렬화/역직렬화를 하려면 어떻게 쓰면 좋을지
고민해보면 의도와 다르게 동작할 수 있는 부분을 막을 수 있을거에요 🤒

NNIIE replied:
Spring에선 Jackson을 직렬화 라이브러리로 채택하고 있는데 gson을 사용할 경우 두 라이브러리간의 직렬화 방식 차이(null 처리 등)에 의한 사이드 이펙트가 발생할 여지가 있음

f-lab-ben commented on src/main/java/com/gamemarket/common/config/DatabaseConfig.java:
```diff
@@ -0,0 +1,25 @@
+public class DatabaseConfig {
+
+    @Bean
+    @ConfigurationProperties(prefix = "spring.datasource")
```
👍

f-lab-ben commented on src/main/java/com/gamemarket/user/ui/UserController.java:
```diff
@@ -0,0 +1,37 @@
+    @ResponseStatus(HttpStatus.CREATED)
+    @Operation(summary = "sign-up")
+    public void signUp(@RequestBody @Valid final UserSignUpRequest request) {
+        if (userService.validateDuplicateEmail(request.getEmail())) {
```
아주 좋습니다~!

다만 서비스 레이어까지 안 가도 되는 경우에는
service를 굳이 참조해야 할 필요가 없다면 repository 정도로만 참고해도 좋을 것 같아요

지금 service 로직은 거의 repository를 래핑해놓은것 뿐이라 
이렇게 되면 repository에 있는 로직이 그대로 다 service로 넘어가서

예전에 SI나 게임회사에 많던
...Repository
...Service
...RepositoryImpl
...ServiceImpl

이런 별 의미없는 모양새가 될 수도 있어요 ㅎㅎ
당장은 래핑하지 말고 repository를 그대로 쓰되,

service 레이어에서 뭔가 조작을 한다거나, validation이 복잡해진다거나 하면
그때 service의 롤로 가져가는게 좋아보입니다
(단순히 지금은 조회만 하기 때문)

NNIIE replied:
좋습니다 
습관적으로 service layer를 참조하게 끔 작성했습니다..!
한층 깔끔해 진거 같아요 👍

f-lab-ben commented:
두가지 의견 드렸으니 이거 고민해보면 좋을것 같아요 ~!

NNIIE committed:
refactor: 직렬화 라이브러리 변경, controller 에서 repository 직접 참조

NNIIE commented:
@f-lab-ben 
피드백 반영하고 다시 PR 요청 드립니다!
다음 2가지 변경사항에 대해 궁금합니다.

<br>

#### 직렬화 부분을 유틸클래스로 생성
유틸리티 클래스 자체가 객체지향적이지 않지만 저는 이렇게 함으로써
객체지향 vs 객체생성에 대한 비용, 제네릭을 통한 재사용성
이라는 trade-off에서 후자를 더 높게 생각하는데 이건 어떻게 생각하시나요?

#### json parsing 예외발생에 대한 로그는 상세히 기록하고, 클라이언트에겐 최소한의 정보를 전달하게끔 의도함
이런 서버로직에 대한 부분은 클라이언트에게 노출할 필요가 없다고 판단했는데, 어떻게 생각하시나요?

f-lab-ben commented:
@NNIIE 
직렬화 부분은 유틸클래스로 작성해도 무방합니다
왜냐면 이건 객체지향적인 순서로 설계할 필요 없이 어디서든 참조해야 하는 경우가 많아서
현업에서도 그렇게 자주 쓰입니다~!
